onejs/comps/Slider.tsx

```diff
@@ -1,10 +1,71 @@
+import math from "math"
+import { JSX, h } from "preact"
 import { Dom } from "OneJS/Dom"
-import { MouseDownEvent, MouseMoveEvent } from "UnityEngine/UIElements"
-import { h } from "preact"
-import { useEffect, useRef, useState } from "preact/hooks"
 import { Style } from "preact/jsx"
+import { useCallback, useEffect, useRef, useState } from "preact/hooks"
+import { IPointerEvent, MouseDownEvent, MouseMoveEvent, PointerDownEvent, PointerMoveEvent, PointerUpEvent } from "UnityEngine/UIElements"
 
-export type SliderProps = {
+export interface SliderProps extends JSX.VisualElement {
+    min?: number
+    max?: number
+    value?: number
+    onChange?: (value: number) => void
+    trackClass?: string
+    trackStyle?: Style
+    activeTrackClass?: string
+    activeTrackStyle?: Style
+    thumbClass?: string
+    thumbStyle?: Style
+}
+
+export function Slider({ min, max, value, onChange, onPointerDown, onPointerMove, onPointerUp, class: $class, trackClass, trackStyle, activeTrackClass, activeTrackStyle, thumbClass, thumbStyle, ...props }: SliderProps): JSX.Element {
+    const trackRef = useRef<Dom>()
+    const activeTrackRef = useRef<Dom>()
+
+    useEffect(() => {
+        const ratio = math.unlerp(min ?? 0, max ?? 1, value ?? min ?? 0)
+        activeTrackRef.current.style.width = `${Math.round(ratio * 100)}%`
+    }, [min, max, value])
+
+    const handlePointerDown = useCallback((e: PointerDownEvent) => {
+        e.currentTarget.CapturePointer(e.pointerId)
+        handlerPointerEvent(e)
+        onPointerDown?.(e)
+    }, [onPointerDown])
+
+    const handlePointerMove = useCallback((e: PointerMoveEvent) => {
+        if (e.currentTarget.HasPointerCapture(e.pointerId)) {
+            handlerPointerEvent(e)
+        }
+        onPointerMove?.(e)
+    }, [onPointerMove])
+
+    const handlePointerUp = useCallback((e: PointerUpEvent) => {
+        if (e.currentTarget.HasPointerCapture(e.pointerId)) {
+            e.currentTarget.ReleasePointer(e.pointerId)
+        }
+        onPointerUp?.(e)
+    }, [onPointerUp])
+
+    const handlerPointerEvent = useCallback((e: IPointerEvent) => {
+        const width = trackRef.current.ve.layout.width
+        const ratio = math.saturate(e.localPosition.x / width)
+        activeTrackRef.current.style.width = `${Math.round(ratio * 100)}%`
+        onChange?.(math.lerp(min ?? 0, max ?? 0, ratio))
```

Comment:
@CreepGin Looks like we need to use `max ?? 1` instead. However, to reduce the probablity of coding error in the future, we should include this code in the component body, before the `useEffect`:
```ts
  min ??= 0
  max ??= 1
  value ??= min
```
And then we don't have to bother with null-coalescing for the rest of the code. It still works correctly with hooks and dependency list.

Reply:
Oh the closure will mess up the type checking:
![image](https://github.com/DragonGround/ScriptLib/assets/666527/5e149f66-ebbe-4b64-a197-872381a3d227)
So I'll adjust it a little bit:
```tsx
const min = _min ?? 0
const max = _max ?? 1
const value = _value ?? min
```

Reply:
That's weird. I don't get that kind of error on my end.

Reply:
Oh do you have "strictNullChecks" off by any chance?

Reply:
Ah, I have `strictNullChecks` off.